Add Donut

NielsRogge opened the pull request:
# What does this PR do?

This PR adds Donut to the library. Donut is to LayoutLM what T5 is to BERT. :D 

The model is implemented as an instance of our existing `VisionEncoderDecoderModel`.

See also https://github.com/clovaai/donut/issues/10#issue-1324734927

To do:

- [x] move repos to appropriate organization in
- [x] update niels to appropriate organization in `test_modeling_vision_encoder_decoder.py`

Commit:
First draft

Commit:
Improve script

Commit:
Update script

Commit:
Make conversion work

Commit:
Add final_layer_norm attribute to Swin's config

Commit:
Add DonutProcessor

Commit:
Convert more models

Commit:
Improve feature extractor and convert base models

Commit:
Fix bug

Commit:
Improve integration tests

Commit:
Improve integration tests and add model to README

Commit:
Add doc test

Commit:
Add feature extractor to docs

Commit:
Fix integration tests

Commit:
Remove register_buffer

Commit:
Fix toctree and add missing attribute

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18488). All of your documentation changes will be reflected on that endpoint.

sgugger commented on src/transformers/models/donut/feature_extraction_donut.py:
```diff
@@ -0,0 +1,207 @@
+            pad_width = delta_width // 2
+            pad_height = delta_height // 2
+
+        padding = (
+            pad_width,
+            pad_height,
+            delta_width - pad_width,
+            delta_height - pad_height,
+        )
```
```suggestion
        padding = (pad_width, pad_height, delta_width - pad_width, delta_height - pad_height)
```
This fits in one line.

sgugger commented on src/transformers/models/swin/configuration_swin.py:
```diff
@@ -75,6 +75,8 @@ class SwinConfig(PretrainedConfig):
             The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
         layer_norm_eps (`float`, *optional*, defaults to 1e-12):
             The epsilon used by the layer normalization layers.
+        add_final_layer_norm (`bool`, *optional*, defaults to True):
+            Whether or not to add a final layer normalization layer after the stages.
```
No we don't add those kind of parameters to configs, this means a new architecture should be created.

NielsRogge replied:
This is resolved as there's now a separate `DonutSwinModel`.

sgugger commented on src/transformers/models/donut/processing_donut.py:
```diff
@@ -0,0 +1,75 @@
+        """
+        return self.tokenizer.decode(*args, **kwargs)
+
+    @contextmanager
+    def as_target_processor(self):
+        """
+        Temporarily sets the tokenizer for processing the input. Useful for encoding the labels when fine-tuning Donut.
+        """
+        self.current_processor = self.tokenizer
+        yield
+        self.current_processor = self.feature_extractor
```
This method is deprecated, there should be a warning here.

NielsRogge replied:
Shouldn't we directly not support it for this model?

sgugger replied:
The deprecation was just introduced after the last release, so I feel we should still support it here to avoid surprise users., but for the one that come after next release, we can avoid introducing it indeed.

sgugger commented on src/transformers/models/donut/processing_donut.py:
```diff
@@ -0,0 +1,75 @@
+        super().__init__(feature_extractor, tokenizer)
+        self.current_processor = self.feature_extractor
+
+    def __call__(self, *args, **kwargs):
+        """
+        When used in normal mode, this method forwards all its arguments to AutoFeatureExtractor's
+        [`~AutoFeatureExtractor.__call__`] and returns its output. If used in the context
+        [`~DonutProcessor.as_target_processor`] this method forwards all its arguments to DonutTokenizer's
+        [`~DonutTokenizer.__call__`]. Please refer to the doctsring of the above two methods for more information.
+        """
+        return self.current_processor(*args, **kwargs)
```
The call should now handle texts and images inputs directly, the docstring should be updated. Look at trocr for an example.

sgugger reviewed (commented):
Thanks for adding this model but:
1. we do not accept configuration parameters to switch layernorm on and off as it hurts readability of the original model. Thus a new model file should be created here.
2. the processor shouldn't implement `as_target_processor` and rely on the new API instead.

Commit:
Add DonutSwin

Commit:
Make conversion script work

Commit:
Improve conversion script

Commit:
Address comment

Commit:
Fix bug

Commit:
Fix another bug

NielsRogge commented:
I have implemented a new DonutSwinModel, that copies everything of SwinModel, except the final layer norm. I've added it in a file called `modeling_donut_swin.py` (and corresponding `configuration_donut_swin.py`), is this ok? Since these only implement the configuration and model of the encoder part (Swin Transformer). For the decoder, BART is leveraged.

NielsRogge commented on docs/source/en/model_doc/donut.mdx:
```diff
@@ -0,0 +1,111 @@
+
+>>> pixel_values = processor(image, return_tensors="pt").pixel_values 
+
+>>> outputs = model.generate(pixel_values.to(device),
+...                          decoder_input_ids=decoder_input_ids.to(device),
+...                          max_length=model.decoder.config.max_position_embeddings,
+...                          early_stopping=True,
+...                          pad_token_id=processor.tokenizer.pad_token_id,
+...                          eos_token_id=processor.tokenizer.eos_token_id,
+...                          use_cache=True,
+...                          num_beams=1,
+...                          bad_words_ids=[[processor.tokenizer.unk_token_id]],
+...                          return_dict_in_generate=True)
```
Can you clarify what you mean by "better"?

sgugger replied:
Run make style after the fix above and you will see.

Commit:
Remove deprecated method from docs

Commit:
Make Swin and Swinv2 untouched

sgugger commented on docs/source/en/model_doc/donut.mdx:
````diff
@@ -0,0 +1,118 @@
+
+- Step-by-step Document Visual Question Answering (DocVQA)
+
+``` py
````
````suggestion
```py
````
The doc styler is not formatting this code example because the `py` is not adjacent to the triple quotes.

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -56,6 +56,7 @@
         ("deit", "DeiTConfig"),
         ("detr", "DetrConfig"),
         ("distilbert", "DistilBertConfig"),
+        ("donut", "DonutSwinConfig"),
```
Not sure we should have this here, as it does not instantiate a Donut model if I understand correctly. For clip for instance, the `"clip"` model types goes with the whole `ClipModel`, not the text/vision part. Maybe have a `donut-swin` model typle like we did for `Data2Vec`?

NielsRogge replied:
Yes, although I saw that for TrOCR, we use "trocr", even though `modeling_trocr`.py and `configuration_trocr`.py only define the model and configuration of the decoder (= text part).

NielsRogge replied:
When I try to do this, `make fixup` and `make fix-copies` complain with the following error:
```
(env) nielsrogge@Nielss-MacBook-Pro transformers % make fix-copies
python utils/check_copies.py --fix_and_overwrite
Traceback (most recent call last):
  File "/Users/nielsrogge/Documents/python_projecten/transformers/src/transformers/utils/import_utils.py", line 1023, in _get_module
    return importlib.import_module("." + module_name, self.__name__)
  File "/opt/homebrew/Cellar/python@3.9/3.9.12/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/Users/nielsrogge/Documents/python_projecten/transformers/src/transformers/models/auto/configuration_auto.py", line 622, in <module>
    class AutoConfig:
  File "/Users/nielsrogge/Documents/python_projecten/transformers/src/transformers/models/auto/configuration_auto.py", line 647, in AutoConfig
    def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
  File "/Users/nielsrogge/Documents/python_projecten/transformers/src/transformers/models/auto/configuration_auto.py", line 609, in docstring_decorator
    lines[i] = _list_model_options(indent, config_to_class=config_to_class, use_model_types=use_model_types)
  File "/Users/nielsrogge/Documents/python_projecten/transformers/src/transformers/models/auto/configuration_auto.py", line 577, in _list_model_options
    lines = [
  File "/Users/nielsrogge/Documents/python_projecten/transformers/src/transformers/models/auto/configuration_auto.py", line 578, in <listcomp>
    f"{indent}- **{model_type}** -- {model_type_to_name[model_type]} ({MODEL_NAMES_MAPPING[model_type]} model)"
KeyError: 'donut-swin'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/nielsrogge/Documents/python_projecten/transformers/utils/check_copies.py", line 549, in <module>
    check_readme(args.fix_and_overwrite)
  File "/Users/nielsrogge/Documents/python_projecten/transformers/utils/check_copies.py", line 512, in check_readme
    model_names_mapping = transformers_module.models.auto.configuration_auto.MODEL_NAMES_MAPPING
  File "/Users/nielsrogge/Documents/python_projecten/transformers/src/transformers/utils/import_utils.py", line 1011, in __getattr__
    value = self._get_module(name)
  File "/Users/nielsrogge/Documents/python_projecten/transformers/src/transformers/utils/import_utils.py", line 1025, in _get_module
    raise RuntimeError(
RuntimeError: Failed to import transformers.models.auto.configuration_auto because of the following error (look up to see its traceback):
'donut-swin'
```

NielsRogge replied:
Updated to `donut-swin`, similar to `Data2Vec`

sgugger commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -56,6 +56,7 @@
         ("deit", "DeiTModel"),
         ("detr", "DetrModel"),
         ("distilbert", "DistilBertModel"),
+        ("donut", "DonutSwinModel"),
```
Same comment here.

NielsRogge replied:
Updated to `donut-swin`

Commit:
Fix code examples

Commit:
Fix processor

Commit:
Update model_type to donut-swin

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -305,7 +306,7 @@
         ("distilbert", "DistilBERT"),
         ("dit", "DiT"),
-        ("donut", "Donut"),
+        ("donut-swin", "Donut"),
```
This one should be `"DonutSwin"`, you can leave another one for the full model (`("donut", "Donut")`).

NielsRogge replied:
Are you sure here? Because I see other models like ("fsmt", "FairSeq Machine-Translation") => isn't this the place where we define what appears in the big table of tasks?

NielsRogge replied:
Ok, so when I add ("donut", "Donut") and ("donut-swin", "Donut") here, the big table gets updated to DonutSwin (which should actually be Donut)

NielsRogge replied:
Gently pinging @sgugger here, as it's my last question before this PR can be merged.

WaterKnight1998 commented on docs/source/en/model_doc/donut.mdx:
```diff
@@ -0,0 +1,169 @@
+>>> sequence = sequence.replace(processor.tokenizer.eos_token, "").replace(processor.tokenizer.pad_token, "")
+>>> sequence = re.sub(r"<.*?>", "", sequence, count=1).strip()  # remove first task start token
+>>> print(sequence)
+<s_question> When is the coffee break?</s_question><s_answer> 11-14 to 11:39 a.m.</s_answer>
```
Are you going to implement any method for converting this outputs into json?

NielsRogge replied:
Ok, I've added it!

alaradirik commented on src/transformers/models/donut/feature_extraction_donut.py:
```diff
@@ -0,0 +1,202 @@
+        self.image_mean = image_mean if image_mean is not None else IMAGENET_STANDARD_MEAN
+        self.image_std = image_std if image_std is not None else IMAGENET_STANDARD_STD
+
+    def rotate(self, image, size):
```
I think it'd be better to move this method to `image_utils`, the utility methods in this file can be used for other common tasks (e.g. preprocessing and inference for pose, keypoint estimation models)

NielsRogge replied:
I've updated the code to move `rotate` to `image_utils.py`. 

However, I still include a `rotate_image` method in DonutFeatureExtractor to check for the size.

Is this ok for you?

alaradirik replied:
Great, that makes sense!

alaradirik commented on src/transformers/models/donut/feature_extraction_donut.py:
```diff
@@ -0,0 +1,202 @@
+
+        return image
+
+    def resize_and_thumbnail(self, image, size, resample):
```
Please see the comment above

NielsRogge replied:
If've added a `do_thumbnail` method, for now this is only defined in `DonutFeatureExtractor`.

alaradirik commented on src/transformers/models/donut/feature_extraction_donut.py:
```diff
@@ -0,0 +1,202 @@
+
+        return image
+
+    def pad(self, image: Image.Image, size: Tuple[int, int], random_padding: bool = False) -> Image.Image:
```
Please see the comment above, this might be better suited for the `image_utils` file

alaradirik commented on src/transformers/models/donut/modeling_donut_swin.py:
```diff
@@ -0,0 +1,941 @@
+
+            Attentions weights after the attention softmax, used to compute the weighted average in the self-attention
+            heads.
+        reshaped_hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each stage) of
+            shape `(batch_size, hidden_size, height, width)`.
```
I think the `reshaped_hidden_states` argument might be a bit redundant, do we really need it?

NielsRogge replied:
These were added to the original Swin Transformer, to make sure the model can be used as a backbone (to return hidden states of shape `(batch_size, num_channels, height, width)` à la convolutional models rather than (batch_size, sequence_length, hidden_size) à la Transformer models.

However, for Donut these may not be necessary, so I can remove them if required.

alaradirik commented on src/transformers/models/donut/modeling_donut_swin.py:
```diff
@@ -0,0 +1,941 @@
+    pooler_output: Optional[torch.FloatTensor] = None
+    hidden_states: Optional[Tuple[torch.FloatTensor]] = None
+    attentions: Optional[Tuple[torch.FloatTensor]] = None
+    reshaped_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
```
Please see the comment above

alaradirik reviewed (commented):
Great work! I think the PR is in good shape but I have some general comments about the structuring.

It'd be great to move the utility methods in `feature_extraction_donut.py` to `image_utils` and also to rethink the naming conventions in test_modeling_vision_encoder_decoder.py if we want to add new models with different base models.

Commit:
Add feature extractor tests, add token2json method, improve feature extractor

Commit:
Fix failing tests, remove integration test

Commit:
Add do_thumbnail for consistency

Commit:
Improve code examples

Commit:
Add code example for document parsing

Commit:
Add DonutSwin to MODEL_NAMES_MAPPING

Commit:
Add model to appropriate place in toctree

Commit:
Update namespace to appropriate organization

gwkrsrch commented on src/transformers/models/donut/convert_donut_to_pytorch.py:
```diff
@@ -0,0 +1,234 @@
+    ]:
+        task_prompt = "<s_cord>"
+    elif model_name == "naver-clova-ix/donut-base-finetuned-cord-v2":
+        task_prompt = "s_cord-v2>"
```
@NielsRogge I found a typo here, "s_cord-v2>"->"<s_cord-v2>".
